core/internal/wbapi/readrunhistoryhandler.go

```diff
@@ -0,0 +1,179 @@
+	spb "github.com/wandb/wandb/core/pkg/service_go_proto"
+)
+
+// RunHistoryAPIHandler handles api requests
+// related to reading a run's history.
+type RunHistoryAPIHandler struct {
+	// currentRequestId is the id of the last scan init request made.
+	//
+	// It is used to provide a unique id for each scan request
+	// and track the associated history reader.
+	currentRequestId atomic.Int32
+
+	// scanHistoryReaders is a map of request ids to history readers.
+	//
+	// It allows us to reuse existing
+	// history readers for subsequent scan requests.
+	scanHistoryReaders map[int32]*runhistoryreader.HistoryReader
+}
+
+func NewRunHistoryAPIHandler() *RunHistoryAPIHandler {
+	return &RunHistoryAPIHandler{
+		currentRequestId:   atomic.Int32{},
+		scanHistoryReaders: make(map[int32]*runhistoryreader.HistoryReader),
+	}
+}
+
+func (f *RunHistoryAPIHandler) HandleRequest(
+	request *spb.ReadRunHistoryRequest,
+) *spb.ApiResponse {
+	switch request.Request.(type) {
+	case *spb.ReadRunHistoryRequest_ScanRunHistoryInit:
+		return f.handleScanRunHistoryInit(request.GetScanRunHistoryInit())
+	case *spb.ReadRunHistoryRequest_ScanRunHistory:
+		return f.handleScanRunHistoryRead(request.GetScanRunHistory())
+	case *spb.ReadRunHistoryRequest_ScanRunHistoryCleanup:
+		f.handleScanRunHistoryCleanup(request.GetScanRunHistoryCleanup())
+	}
+
+	return nil
+}
+
+// handleScanRunHistoryInit handles a request to initialize
+// a scan over a run's history.
+//
+// It creates a new history reader for the run,
+// and caches it for subsequent scan requests.
+// It returns an Id correlating subsequent scan requests
+// with the history reader.
+func (f *RunHistoryAPIHandler) handleScanRunHistoryInit(
+	request *spb.ScanRunHistoryInit,
+) *spb.ApiResponse {
+	requestId := f.currentRequestId.Add(1)
+	requestKeys := request.GetKeys()
+	settings := settings.From(request.GetSettings())
+
+	backend := stream.NewBackend(observability.NewNoOpLogger(), settings)
+	graphqlClient := stream.NewGraphQLClient(
+		backend,
+		settings,
+		&observability.Peeker{},
+		"", /*clientId*/
+	)
+
+	historyReader := runhistoryreader.New(
+		context.Background(),
+		request.Entity,
+		request.Project,
+		request.RunId,
+		graphqlClient,
+		http.DefaultClient,
+		requestKeys,
+	)
+	f.scanHistoryReaders[requestId] = historyReader
+
+	return &spb.ApiResponse{
+		Response: &spb.ApiResponse_ReadRunHistoryResponse{
+			ReadRunHistoryResponse: &spb.ReadRunHistoryResponse{
+				Response: &spb.ReadRunHistoryResponse_ScanRunHistoryInit{
+					ScanRunHistoryInit: &spb.ScanRunHistoryInitResponse{
+						RequestId: requestId,
+					},
+				},
+			},
+		},
+	}
+}
+
+// handleScanRunHistoryRead handles a request to scan
+// over a portion of a run's history.
+func (f *RunHistoryAPIHandler) handleScanRunHistoryRead(
+	request *spb.ScanRunHistory,
+) *spb.ApiResponse {
+	requestId := request.GetRequestId()
+
+	historyReader := f.scanHistoryReaders[requestId]
+
+	if historyReader == nil {
```

Comment:
I'd also defensively check if it exists `historyReader, ok := f.scanHistoryReaders[requestId]`, `if !ok || hr == nil`